crates/rune-macros/src/context.rs

```diff
@@ -113,6 +117,57 @@ pub(crate) struct FieldProtocol {
     custom: Option<syn::Path>,
 }
 
+pub(crate) struct TypeProtocol {
+    protocol: syn::Ident,
+    handler: Option<syn::Path>,
+}
+
+impl TypeProtocol {
+    pub fn expand(&self) -> TokenStream {
+        if let Some(handler) = &self.handler {
+            let protocol = &self.protocol;
+            return quote_spanned! {protocol.span()=>
+                module.associated_function(rune::runtime::Protocol::#protocol, #handler)?;
+            };
+        }
+        match self.protocol.to_string().as_str() {
+            "ADD" => quote_spanned! {self.protocol.span()=>
+                module.associated_function(rune::runtime::Protocol::ADD, |this: Self, other: Self| this + other)?;
+            },
+            "STRING_DISPLAY" => quote_spanned! {self.protocol.span()=>
+                module.associated_function(rune::runtime::Protocol::STRING_DISPLAY, |this: &Self, buf: &mut String| {
+                    use ::core::fmt::Write as _;
+                    ::core::write!(buf, "{this}")
+                })?;
+            },
+            "STRING_DEBUG" => quote_spanned! {self.protocol.span()=>
+                module.associated_function(rune::runtime::Protocol::STRING_DEBUG, |this: &Self, buf: &mut String| {
+                    use ::core::fmt::Write as _;
+                    ::core::write!(buf, "{this:?}")
+                })?;
+            },
+            _ => syn::Error::new_spanned(
+                &self.protocol,
+                format!("Rune protocol `{}` cannot be derived", self.protocol),
+            )
+            .to_compile_error(),
```

Comment:
Make sure to raise this error during parsing instead. I initially missed this, but I want to make sure that error handling is as coherent as possible in that it's only handled in a single place. This results in a `compile_error!` being produces in the middle of the install block which while clever is a bit surprising.